### PR TITLE
H-642: Allow mass creation/deletion of relationships

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

SpiceDB allows to modify (touch, create, delete) relations. For simplicity we only expose a non-batched version yet. This PR improves some DX issues we had previously (discussed offline) and allows multiple tuples to be inserted/deleted at the same time.

## 🔍 What does this change?

- Support preconditions in modification endpoints
- Simplify interface to improve DX and probably compile time
- Allow multiple tuples to be passed in modification endpoints

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

## 🛡 What tests cover this?

The authorization integration tests uses the new batch creation/deletion methods

### CODE REVIEW

File: apps/hash-graph/lib/authorization/src/zanzibar.rs [hunk 1]
```diff
@@ -5,234 +5,223 @@ use std::borrow::Cow;
 
 use serde::{Deserialize, Serialize};
 
-/// The relation or permission of a [`Subject`] to an [`Resource`].
+/// The relation or permission of a [`Resource`] to another [`Resource`].
 pub trait Affiliation<R: Resource + ?Sized>: AsRef<str> {}
 
-/// A computed set of [`Subject`]s for a particular [`Resource`].
+impl<A: Affiliation<R>, R: Resource> Affiliation<R> for &A {}
+// impl<A: Affiliation<R>, R: Resource> Affiliation<&R> for A {}
```
Comment (indietyp) [inline apps/hash-graph/lib/authorization/src/zanzibar.rs:12]: Why is this commented out, and if it is not needed why is it still here?

Comment (TimDiekmann) [reply apps/hash-graph/lib/authorization/src/zanzibar.rs:12]: I forgot to remove that, thank you

File: apps/hash-graph/lib/authorization/src/backend.rs [hunk 2]
```diff
@@ -33,73 +31,58 @@ pub trait AuthorizationApi {
...
 
-    /// Creates a new relation between a [`Subject`] and an [`Resource`] with the specified
-    /// [`Relation`].
+    /// Creates a new relation specified by the [`Tuple`].
     ///
     /// # Errors
     ///
     /// Returns an error if the relation already exists or could not be created.
-    async fn create_relation<R, A, S>(
+    async fn create_relation<'p, 't, T>(
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply apps/hash-graph/lib/authorization/src/backend.rs:39]: Good point, thanks!

File: apps/hash-graph/lib/authorization/src/zanzibar.rs (hunk 1 above)
Comment (TimDiekmann) [reply apps/hash-graph/lib/authorization/src/zanzibar.rs]: Removed in da0abcf0a3792aff1edcd21cd699fa2dbb623886. I expect to remove all reference-based parameters as all types should implement `Clone` anyway and this avoids issues like passing reference to tuples and such.

File: apps/hash-graph/lib/authorization/src/backend.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply apps/hash-graph/lib/authorization/src/backend.rs:39]: If you don't mind I'll do that in my follow up to avoid conflicts.

**Review by indietyp — APPROVED** [review]

**Pull request merged**